File: backend/app/models/provenance.py
"""Provenance models — the spine every fact/claim references (Invariant #3)."""

from __future__ import annotations

from typing import Literal
from uuid import uuid4

from pydantic import BaseModel, Field


def _new_id() -> str:
    return str(uuid4())


class SourceQuery(BaseModel):
    """A record of one external call (or import). Facts/claims FK to this, written in the
    same transaction. ``params`` is a dict here (serialized to JSON on write) and MUST include
    applied expansion bounds for address-scoped capabilities (audit #10)."""

    id: str = Field(default_factory=_new_id)
    connector: str
    capability: str
    endpoint: str
    params: dict | None = None
    requested_at: str
    completed_at: str | None = None
    status: Literal["ok", "error", "partial"] = "ok"
    result_summary: str | None = None
    # raw_response_ref / raw_response_hash are set by the writer from the raw bytes.


class Exhibit(BaseModel):
    id: str = Field(default_factory=_new_id)
    exhibit_type: Literal["screenshot", "file", "export"]
    source: str | None = None
    captured_at: str
    file_ref: str
    content_hash: str
    description: str | None = None

File: frontend/src/provenance.ts
// Provenance drill-through (FN-01): fetch the full `source_query` behind a displayed claim/fact, so an
// investigator can see WHERE it came from — connector, endpoint, params/bounds, retrieval time, and the
// raw-response hash — in one interaction. Read-only; the never-collapse model keeps each source separate
// (Invariant #4), and this just exposes the query each claim already references.

export type SourceQueryProvenance = {
  id: string;
  connector: string;
  capability: string;
  endpoint: string;
  params: Record<string, unknown> | string | null;
  requested_at: string;
  completed_at: string | null;
  status: string;
  raw_response_ref: string | null;
  raw_response_hash: string | null;
  result_summary: string | null;
};

export function fetchSourceQuery(id: string): Promise<SourceQueryProvenance> {
  return fetch(`/api/source_query/${encodeURIComponent(id)}`).then((r) => {
    if (!r.ok) throw new Error(`HTTP ${r.status}`);
    return r.json();
  });
}
